core/src/main/java/org/apache/iceberg/rest/HTTPClient.java

```diff
@@ -214,92 +225,63 @@ private static void throwFailure(
-        (path.startsWith("https://") || path.startsWith("http://"))
-            ? path
-            : String.format("%s/%s", uri, path);
-    try {
-      URIBuilder builder = new URIBuilder(fullPath);
-      if (params != null) {
-        params.forEach(builder::addParameter);
-      }
-      return builder.build();
-    } catch (URISyntaxException e) {
-      throw new RESTException(
-          "Failed to create request URI from base %s, params %s", fullPath, params);
-    }
-  }
-
-  /**
-   * Method to execute an HTTP request and process the corresponding response.
-   *
-   * @param method - HTTP method, such as GET, POST, HEAD, etc.
-   * @param queryParams - A map of query parameters
-   * @param path - URI to send the request to
-   * @param requestBody - Content to place in the request body
-   * @param responseType - Class of the Response type. Needs to have serializer registered with
-   *     ObjectMapper
-   * @param errorHandler - Error handler delegated for HTTP responses which handles server error
-   *     responses
-   * @param <T> - Class type of the response for deserialization. Must be registered with the
-   *     ObjectMapper.
-   * @return The response entity, parsed and converted to its type T
-   */
-  private <T> T execute(
-      Method method,
+  @Override
+  protected HTTPRequest buildRequest(
+      HTTPMethod method,
       String path,
       Map<String, String> queryParams,
-      Object requestBody,
-      Class<T> responseType,
       Map<String, String> headers,
-      Consumer<ErrorResponse> errorHandler) {
-    return execute(
-        method, path, queryParams, requestBody, responseType, headers, errorHandler, h -> {});
+      Object body) {
+
+    ImmutableHTTPRequest.Builder builder =
+        ImmutableHTTPRequest.builder()
+            .baseUri(baseUri)
+            .mapper(mapper)
+            .method(method)
+            .path(path)
+            .body(body)
+            .queryParameters(queryParams == null ? Map.of() : queryParams);
+
+    Map<String, String> allHeaders = Maps.newLinkedHashMap();
+    if (headers != null) {
+      allHeaders.putAll(headers);
+    }
+
+    allHeaders.putIfAbsent(HttpHeaders.ACCEPT, ContentType.APPLICATION_JSON.getMimeType());
+
+    // Many systems require that content type is set regardless and will fail,
+    // even on an empty bodied request.
+    // Encode maps as form data (application/x-www-form-urlencoded),
+    // and other requests are assumed to contain JSON bodies (application/json).
+    ContentType mimeType =
+        body instanceof Map
+            ? ContentType.APPLICATION_FORM_URLENCODED
+            : ContentType.APPLICATION_JSON;
+    allHeaders.putIfAbsent(HttpHeaders.CONTENT_TYPE, mimeType.getMimeType());
+
+    // Apply base headers now to mimic the behavior of
+    // org.apache.hc.client5.http.protocol.RequestDefaultHeaders
+    // We want these headers applied *before* the AuthSession authenticates the request.
+    if (baseHeaders != null) {
```

Comment:
Doesn't this mean that `baseHeaders` will override the provided specific headers, the mime type headers and accept headers?  It feels like this should go first.

Reply:
This code mimics the behavior pre-auth manager: `baseHaders` was fed into `RequestDefaultHeaders`, which is applied quite at the end of the request processing, after other headers are set. However, `RequestDefaultHeaders` never overwrites a header that was previously set by client code.

Thus here we have the following order of precedence to achieve the same result:

- Headers passed to the method call directly are applied first;
- Then, `baseHeaders` are applied, but using `putIfAbsent` so that they do not overwrite any headers;
- And finally (this is new), `authSession` applies its headers. All implementations of `AuthSession` use `putIfAbsent` as well, so they do not overwrite any existing headers.